src/main/scala/scalismo/registration/ImageMetric.scala

```diff
@@ -0,0 +1,120 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scalismo.registration
+
+import scalismo.common.Domain
+import scalismo.geometry._
+import scalismo.image.{ DifferentiableScalarImage, ScalarImage }
+import scalismo.numerics.{ Sampler, Integrator }
+import scalismo.registration.ImageMetric.ValueAndDerivative
+import scalismo.utils.Random
+
+import scala.language.higherKinds
+import breeze.linalg.DenseVector
+
+trait ImageMetric[D <: Dim] {
+
+  /**
+   * Computes the metric value M[img1 . transformation, img2] (where a . b denotes composition)
+   */
+  def value(img1: ScalarImage[D], img2: ScalarImage[D], transform: Transformation[D]): Double
+  implicit def ndSpace: NDSpace[D]
+
+  /**
+   * Computes the derivative of the metric value M[img1 . transformation, img2], with respect to
+   * the transform parameters, for the given transformation space and the transformation defined
+   * at the current parameter values.
+   */
+  def takeDerivativeWRTToTransform(im1: ScalarImage[D], img2: DifferentiableScalarImage[D],
+    transformationSpace: TransformationSpace[D], params: DenseVector[Double]): DenseVector[Double]
+
+  /**
+   * Computes value and derivative in one go.
+   */
+  def computeValueAndDerivative(img1: ScalarImage[D], img2: DifferentiableScalarImage[D],
+    transformationSpace: TransformationSpace[D], params: DenseVector[Double]): ValueAndDerivative
+}
+
+object ImageMetric {
+  case class ValueAndDerivative(value: Double, derivative: DenseVector[Double])
+}
+
+case class MeanSquaresMetric[D <: Dim: NDSpace](sampler: Sampler[D]) extends ImageMetric[D] {
+
+  override val ndSpace = implicitly[NDSpace[D]]
+
+  def value(fixedImage: ScalarImage[D], movingImage: ScalarImage[D], transform: Transformation[D]) = {
+    computeValue(fixedImage, movingImage, transform, Integrator(sampler))
+  }
+
+  // compute the derivative of the cost function
+  def takeDerivativeWRTToTransform(fixedImage: ScalarImage[D], movingImage: DifferentiableScalarImage[D],
+    transformationSpace: TransformationSpace[D], params: DenseVector[Double]): DenseVector[Double] = {
+    computeDerivative(fixedImage, movingImage, transformationSpace, params, Integrator(sampler))
+  }
+
+  override def computeValueAndDerivative(fixedImage: ScalarImage[D], movingImage: DifferentiableScalarImage[D],
+    transformationSpace: TransformationSpace[D], params: DenseVector[Double]): ValueAndDerivative = {
+
+    // We create a new sampler, which always returns the same points. In this way we can make sure that the
+    // same sample points are used for computing the value and the derivative
+    val sampleOnceSampler = new Sampler[D] {
```

Comment:
@Ghazi-Bouabene  This is intentional. The idea here is that the same sample is used for computing the value and the derivative.  But in every call of ```valueAndDerivative``` a new sample should be chosen. The other behavior can be achieved by passing a sampler to the metric, which always samples the same points.